keps/0029-operator-dependencies.md

```diff
@@ -0,0 +1,203 @@
+```
+
+In the first step we build a dependency graph. A set of all graph vertices (which are task-operators) `S` is defined as `S = {AA, BB, CC, EE, GG}`. A transitive relationship `R` between the vertices is defined as `(a, b) ∈ S` meaning that _`a` needs `b` deployed first_. The transitive relationship for the above example is: `R = { (AA,BB), (AA,CC), (BB,EE), (BB,GG) }`. The resulting topological order `O` is therefor `O = (EE, GG, BB, CC, AA)` which has no cycles.
+
+The instance controller (IC) then traverses the dependency graph in the evaluation order `O` and executes each vertex. Practically this means creating a corresponding `Instance` resource (`Operator` and `OperatorVersion` already exist) and wait for it to become healthy, meaning that its `deploy` plan was executed successfully. KUDO manager already has [health check](https://github.com/kudobuilder/kudo/blob/master/pkg/engine/health/health.go#L78) for `Instance` resources implemented. We would additionally add the top-level `Instance` reference (e.g. `AA`) to the `ownerReferences` list of each dependency `Instance` (e.g. `BB`). This would help with determining which `Instance` belongs to which operator and additionally help us with [operator uninstalling](#uninstalling).
+
+The status of the execution can be seen as usual as part of the `Instance.Status`. We could additionally forward the status of a dependency `Instance` to the top-level `Instance.Status` to simplify the overview.
+
+##### Dependencies Parameterization
+
+First, we need to provide parameters to different operators separately. For individual parameters (`-p` option) we can use namespaced names e.g. `-p <operatorName>.<key>=<value>`. For the parameter files (`--parameterFile` option) we could extend `parameterFile` schema (which is currently a simple map) with top-level fields e.g.:
+
+```yaml
+apiVersion: v1beta1
+operator: zookeeper
+parameters:
+  foo: bar # other parameterFile keys and values
+```
+
+Second, parameters are normally parsed to the CLI and persisted in the `Instance` resource. Since the creation of the instances is delayed we need a place to store the parameters until then. We could keep them in the task definition e.g.:
+
+```yaml
+tasks:
+- name: demo
+  kind: Operator
+  spec:
+    ...
+    parameters: # optional, a map of parameter keys and values constructed by expanding the parameterFiles along with individual parameters
+
+```
+
+KUDO manager would then read them and populate `Instance.Spec.Parameters` as usual. The drawback here is that this would likely stretch currently slim task definitions.
+Alternatively, we could introduce a new field e.g. `OperatorVersionSpec.InstallationParameters` alongside with the existing [Parameters](https://github.com/kudobuilder/kudo/blob/master/pkg/apis/kudo/v1beta1/operatorversion_types.go#L35). Both approaches contradict our current data model where such parameters would exist in the `Instance` resource.
```

Comment:
Saving this in the OperatorVersionSpec would also clash with - as mentioned above - multiple instances that need to be installed.

Reply:
I've added a clarification in the KEP 👍